File: Editorial/0091_Decode_Ways/CSharp/1_Recursive_Approach_with_Memoization.cs
public class Solution {
    private Dictionary<int, int> memo = new Dictionary<int, int>();

    private int RecursiveWithMemo(int index, string s) {
        if (memo.ContainsKey(index)) {
            return memo[index];
        }

        if (index == s.Length) {
            return 1;
        }

        if (s[index] == '0') {
            return 0;
        }

        if (index == s.Length - 1) {
            return 1;
        }

        int ans = RecursiveWithMemo(index + 1, s);
        if (int.Parse(s.Substring(index, 2)) <= 26) {
            ans += RecursiveWithMemo(index + 2, s);
        }

        memo[index] = ans;
        return ans;
    }

    public int NumDecodings(string s) {
        return RecursiveWithMemo(0, s);
    }
}

File: Editorial/0000-0100/0091_Decode_Ways/JavaScript/1_Recursive_Approach_with_Memoization.js
var numDecodings = function (s) {
    let memo = new Object();
    return recursiveWithMemo(0, s, memo);
};
const recursiveWithMemo = (index, str, memo) => {
    if (memo.hasOwnProperty(index)) {
        return memo[index];
    }
    if (index == str.length) {
        return 1;
    }
    if (str.charAt(index) == "0") {
        return 0;
    }
    if (index == str.length - 1) {
        return 1;
    }
    let ans = recursiveWithMemo(index + 1, str, memo);
    if (parseInt(str.substring(index, index + 2)) <= 26) {
        ans += recursiveWithMemo(index + 2, str, memo);
    }
    memo[index] = ans;
    return ans;
};
